Re-write crate level API

tcharding opened the pull request:
Now includes updates to fuzzing and is on top of #130 (bug fix).

Now we have all the new `primitives` pieces in place we can re-write the public API to take advantage of them - WIN!

Close: #126

tcharding commented:
This PR more or less follows the proposal in #126 but I'm not auto-closing it because it does not include the suggested checksum example.

> If you want to define your own checksum you should use the Checksum trait. We should provide an example, maybe of the BIP-93 checksum.

tcharding commented:
This might need splitting up into separate PRs now because I've made some changes to the HRP constants to make them more readable.

I've audited the two top level APIs (lib.rs and segwit.rs) and I'm happy they are uniform. Open to naming improvements still of course.

I've written docs and examples, still TODO is the custom checksum doc.

Note please the new `crate::hrp` module that is soley to improve import ergonomics because I noticed we use `hrp::BC` any time we use the segwit API but one used to have to reach into `primitives` to get it.

tcharding commented:
Ok, if CI goes green this is ready for review please @apoelstra, I edited the "what I did" post above, so in case you read a stale version please not this now adds `Codex32`. If you are more-or-less happy with the current state I can spilt the PR up. I'll ping you @clarkmoody when we are at a "final" re-viewable state so you don't have to follow along if you don't want to.

tcharding commented:
FYI to check the embedded tests locally one must pin with `cargo update -p proc-macro2 --precise 1.0.60`. I can't remember how to run the crate, using `run` causes a core dump.

tcharding commented:
In case its useful to you @clarkmoody I'm building the docs locally to review them using

`RUSTDOCFLAGS="--cfg docsrs" cargo +nightly rustdoc --all-features -- -D rustdoc::broken-intra-doc-links`

apoelstra commented on embedded/no-allocator/src/main.rs:
```diff
@@ -20,20 +20,21 @@ use panic_halt as _;
-    let hrp = Hrp::parse("bech32").unwrap();
-
-    bech32::encode_to_fmt_anycase(&mut encoded, hrp, &base32, Variant::Bech32).unwrap().unwrap();
+    bech32::encode_to_fmt::<Bech32, ArrayString<30>>(&mut encoded, hrp, &data)
```
In 4da0a7b9784a7fe92db39a7d645d4bee97999d4d:

Could use `<Bech32, _>` rather than repeating `ArrayString<30>`.

tcharding replied:
Done, thanks.

apoelstra commented on src/primitives/mod.rs:
```diff
@@ -26,6 +26,11 @@ pub enum Bech32 {}
+/// The codex32 checksum algorithm, defined in [BIP-93].
+/// [BIP-350]: <https://github.com/bitcoin/bips/blob/master/bip-0093.mediawiki>
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Codex32 {}
```
In 4da0a7b9784a7fe92db39a7d645d4bee97999d4d:

This can live in a doccomment or something but we probably shouldn't export it from this library.

tcharding replied:
Done.

clarkmoody commented on src/lib.rs:
```diff
@@ -30,1209 +106,244 @@ extern crate core;
+pub use crate::primitives::{Bech32, Bech32m, NoChecksum};
 
 mod error;
+/// Re-exports the hrp types from [`primitives::hrp`] to make importing ergonomic for the top level APIs.
+pub mod hrp;
```
Love the re-exports to the top-level `hrp` module.

Do the doc comments need to point out that these are re-exports?

apoelstra replied:
There is a stupid compiler warning now if re-exports don't have doccomments. Since they appear in the docs under `Re-exports` with a link to their definition, this isn't really sensible, so we've been putting "placeholder comments" like this in rust-bitcoin.

tcharding replied:
Commented on this below.

clarkmoody commented on src/primitives/mod.rs:
```diff
@@ -26,6 +26,11 @@ pub enum Bech32 {}
 pub enum Bech32m {}
 
+/// The codex32 checksum algorithm, defined in [BIP-93].
+/// [BIP-350]: <https://github.com/bitcoin/bips/blob/master/bip-0093.mediawiki>
```
Should this be `[BIP-93]` here?

tcharding replied:
Thanks, odd that `-D rustdoc::broken-intra-doc-links` didn't catch that. I've moved the codex32 stuff to a rustdoc comment and remove the link altogether.

clarkmoody commented on src/lib.rs:
```diff
@@ -30,1209 +106,244 @@ extern crate core;
+pub mod hrp;
+/// All the primitive types and functionality used in encoding and decoding.
 pub mod primitives;
+/// API for encoding and decoding segwit addresses.
```
In the docs, this comment collides with the existing module-level docs in `segwit.rs`, giving us:

> API for encoding and decoding segwit addresses. Segregated Witness API - enables typical usage for encoding and decoding segwit addresses.

We should remove one of the comments to avoid the redundant sentence.

tcharding replied:
Ha, I saw this once and didn't fix it. Thanks for taking the time to build the docs and check them!

I've removed the docs from `rust-bech32/src/hrp.rs` and clippy seems to be ok with it, although I'm unsure why. Like @apoelstra said above we saw a new lint trigger for this situation in rust-bitcoin, will see how it goes in CI.

clarkmoody reviewed (commented):
Nice top-level examples in the doc code!

tcharding commented:
> In [4da0a7b](https://github.com/rust-bitcoin/rust-bech32/commit/4da0a7b9784a7fe92db39a7d645d4bee97999d4d):
> 
> * I'm a bit confused why `decode` and `decode_bech32m`/`decode_bech32`.

Ok, I removed them and added more docs about decoding with the checksum algorithm to the `decode` docs.

> * Could you add an analogous `encode_to_write` / `decode_to_read` pair to `encode_to_fmt`? Then no-alloc users could encode the string either to an `io::Write` or `fmt::Write`. (Annoyingly this will require a std gate, but that's ok, it would still be useful I think.)

Done!

> Otherwise this looks great!

Thanks man, this force push was the result of a pretty wild session, you can wait to review until I've had a chance to re-look at it with fresh eyes to catch my own mistakes if you like.

apoelstra commented on src/segwit.rs:
```diff
@@ -100,6 +102,19 @@ pub fn encode(
+) -> Result<(Hrp, Fe32, Vec<u8>), DecodeFromReaderError> {
+    let mut s = String::new();
+    r.read_to_string(&mut s)?;
+    Ok(decode(&s)?)
```
In 68c4dfd2666ff390cc10d14adc7d5302667e3f20:

I'll go ahead and ACK this since it's not API-visible, but the string allocation (and UTF-8 checks and everything that comes with it) are all unnecessary here. You should be able to just use the `bytes()` iterator on the `Read`, adapt it wiht `char::from` to get characters, then use the iterator API.

tcharding replied:
I like it! I'll have go.

tcharding replied:
I had a bit of a play with this, unless I'm missing something:
- The iterator API (you mean `iter.rs`, right?) is for encoding not decoding
- The API in `decode.rs` could be changed to enable decoding from a reader but its non-trivial because all the decoding types (eg, `UncheckedHrpstring`) have lifetimes that expect a string somewhere to reference.

We could read into a buffer because a bech32 string is max 90 chars, then parse it out however we like.

Or am I totally missing something?

apoelstra replied:
No, you may be correct. Several iterations ago I had an idea for decoding through the iterator API and I guess that got dropped.

I think it's fine to just drop this function then. I don't have a clear usecase for it and it doesn't even work with nostd.

tcharding replied:
Removed, both in `lib.rs` and `segwit.rs`

apoelstra reviewed (approved):
ACK 68c4dfd2666ff390cc10d14adc7d5302667e3f20

apoelstra reviewed (approved):
ACK 9a90d64b6f182d31c80c67d80c831862b772c390

tcharding commented:
Rebaes only, no other changes.

apoelstra reviewed (approved):
ACK 79918f664785ce6b4589842713898db9ea93259b

clarkmoody reviewed (approved):
ACK 79918f6
